lib/rhc/commands/port_forward.rb

```diff
@@ -73,55 +74,77 @@ class PortForward < Base
     option ["-g", "--gear ID"], "Gear ID you are port forwarding to (optional)"
     option ["-s", "--service [SERVICE,]"], "A CSV list of services to port forward (optional)"
     def run(app)
-      rest_app = find_app
-      ssh_uri = URI.parse(options.gear ? rest_app.gear_ssh_url(options.gear) : rest_app.ssh_url)
-
-      say "Using #{ssh_uri}..." if options.debug
-
+      ssh_executable = check_ssh_executable! options.ssh
       forwarding_specs = []
 
       begin
-        say "Checking available ports ... "
-
-        Net::SSH.start(ssh_uri.host, ssh_uri.user) do |ssh|
-          # If a specific gear is targeted, do not include remote (e.g. database) ports
-          list_ports_cmd = "rhc-list-ports#{options.gear ? ' --exclude-remote' : ''}"
-          ssh.exec! list_ports_cmd do |channel, stream, data|
-            if stream == :stderr
-              data.each_line do |line|
-                line.chomp!
-                # FIXME: This is really brittle; there must be a better way
-                # for the server to tell us that permission (what permission?)
-                # is denied.
-                raise RHC::PermissionDeniedException.new "Permission denied." if line =~ /permission denied/i
-                # ...and also which services are available for the application
-                # for us to forward ports for.
-                if line =~ /\A\s*(\S+) -> #{HOST_AND_PORT}\z/ and (options.service.nil? or options.service.empty? or options.service.split(',').include? $1)
-                  debug fs = ForwardingSpec.new($1, $2, $3.to_i)
-                  forwarding_specs << fs
-                else
-                  debug line
-                end
+        rest_app = find_app
+        ssh_uri = URI.parse(options.gear ? rest_app.gear_ssh_url(options.gear) : rest_app.ssh_url)
+        debug "Using #{ssh_uri}..."
 
+        output = ""
+        say "Checking available ports ... "
+        # Only use Net::SSH if no ssh executable is specified
+        if !options.ssh
+          Net::SSH.start(ssh_uri.host, ssh_uri.user) do |ssh|
+            # If a specific gear is targeted, do not include remote (e.g. database) ports
+            list_ports_cmd = "rhc-list-ports#{options.gear ? ' --exclude-remote' : ''}"
+            ssh.exec! list_ports_cmd do |channel, stream, data|
+              if stream == :stderr
+                output = data
```

Comment:
I think you want `output << data`.

"the block is called for each data and extended data packet"

https://net-ssh.github.io/ssh/v2/api/classes/Net/SSH/Connection/Session.html#M000093